File: android/app/src/main/java/com/mototour/app/data/RouteCalculator.kt
package com.mototour.app.data

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

object RouteCalculator {

    private const val OSRM_BASE = "https://router.project-osrm.org/route/v1/driving"

    /**
     * Calculate a road-following route between the given waypoints using OSRM.
     * Returns the route as a list of [GpxPoint], or null if the request fails.
     */
    suspend fun calculateRoute(waypoints: List<GpxPoint>): List<GpxPoint>? {
        if (waypoints.size < 2) return null

        return withContext(Dispatchers.IO) {
            try {
                val coords = waypoints.joinToString(";") { "${it.lon},${it.lat}" }
                val url = URL("$OSRM_BASE/$coords?overview=full&geometries=polyline")
                val conn = url.openConnection() as HttpURLConnection
                conn.connectTimeout = 10_000
                conn.readTimeout = 10_000
                conn.setRequestProperty("User-Agent", "MotoTourApp/1.0")

                try {
                    if (conn.responseCode != 200) return@withContext null
                    val body = conn.inputStream.bufferedReader().readText()
                    val json = JSONObject(body)
                    if (json.optString("code") != "Ok") return@withContext null

                    val geometry = json.getJSONArray("routes")
                        .getJSONObject(0)
                        .getString("geometry")

                    decodePolyline(geometry)
                } finally {
                    conn.disconnect()
                }
            } catch (_: Exception) {
                null
            }
        }
    }

    /**
     * Decode a Google encoded polyline string into a list of [GpxPoint].
     * See: https://developers.google.com/maps/documentation/utilities/polylinealgorithm
     */
    private fun decodePolyline(encoded: String): List<GpxPoint> {
        val points = mutableListOf<GpxPoint>()
        var index = 0
        var lat = 0
        var lng = 0

        while (index < encoded.length) {
            // Decode latitude
            var shift = 0
            var result = 0
            var b: Int
            do {
                b = encoded[index++].code - 63
                result = result or ((b and 0x1F) shl shift)
                shift += 5
            } while (b >= 0x20)
            lat += if (result and 1 != 0) (result shr 1).inv() else result shr 1

            // Decode longitude
            shift = 0
            result = 0
            do {
                b = encoded[index++].code - 63
                result = result or ((b and 0x1F) shl shift)
                shift += 5
            } while (b >= 0x20)
            lng += if (result and 1 != 0) (result shr 1).inv() else result shr 1

            points.add(GpxPoint(lat / 1e5, lng / 1e5))
        }
        return points
    }
}

File: web/src/lib/routeCalculator.js
/**
 * Fetch a road-following route from the public OSRM server.
 * Mirrors RouteCalculator.kt exactly.
 *
 * Returns an array of { lat, lon } points, or null on any failure.
 */

const OSRM_BASE = 'https://router.project-osrm.org/route/v1/driving'

export async function fetchOsrmRoute(routePoints) {
  if (!routePoints || routePoints.length < 2) return null
  try {
    const coords = routePoints.map((p) => `${p.lon},${p.lat}`).join(';')
    const url = `${OSRM_BASE}/${coords}?overview=full&geometries=polyline`
    const resp = await fetch(url, { signal: AbortSignal.timeout(10_000) })
    if (!resp.ok) return null
    const json = await resp.json()
    if (json.code !== 'Ok') return null
    return decodePolyline(json.routes[0].geometry)
  } catch {
    return null
  }
}

/**
 * Decode a Google-encoded polyline string into [{ lat, lon }].
 * Direct JS port of decodePolyline() in RouteCalculator.kt.
 */
function decodePolyline(encoded) {
  const points = []
  let index = 0
  let lat = 0
  let lng = 0

  while (index < encoded.length) {
    let shift = 0
    let result = 0
    let b
    do {
      b = encoded.charCodeAt(index++) - 63
      result |= (b & 0x1f) << shift
      shift += 5
    } while (b >= 0x20)
    lat += result & 1 ? ~(result >> 1) : result >> 1

    shift = 0
    result = 0
    do {
      b = encoded.charCodeAt(index++) - 63
      result |= (b & 0x1f) << shift
      shift += 5
    } while (b >= 0x20)
    lng += result & 1 ? ~(result >> 1) : result >> 1

    points.push({ lat: lat / 1e5, lon: lng / 1e5 })
  }
  return points
}
